### PR TITLE
Write a test for slider snaking

### PR DESCRIPTION
**Pull request by LittleEndu**
* [x] Depends on #8486

This PR writes a test for slider snaking. 

Behaviour that is tested both when snaking is enabled and disabled:
* Sliders snake in
* Sliders snake out
* Repeat sliders don't snake out unless all repeats are over
  * Tested for a slider with one repeat
  * Tested for a slider with two repeats 

Also includes tests for reverse arrow behaviour when snaking is enabled.

### CODE REVIEW

File: osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+            {
+                new Slider
+                {
+                    StartTime = 3000,
+                    Position = new Vector2(100, 100),
+                    Path = new SliderPath(PathType.PerfectCurve, new[]
+                    {
+                        Vector2.Zero,
+                        new Vector2(300, 200)
+                    }),
```
Comment (bdach) [inline osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs]: Seeing that we don't really care about position & path in these sliders I'd maybe extract out a helper method and take in `StartTime` and `RepeatCount` to make this a little leaner?

File: osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+            testDoubleRepeating(true);
+            testDoubleRepeating(false);
+        }
+
+        [TestCase(true)]
+        [TestCase(false)]
+        public void TestArrowStays(bool isHit)
+        {
+            var isSame = isHit ? "is same" : "decreased";
+            var enable = isHit ? "enable" : "disable";
```
Comment (bdach) [inline osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs]: These are named super confusingly (having `isSame` not have a value of `"is same"` will trip people up).

File: osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+        [Resolved]
+        private AudioManager audioManager { get; set; }
+
+        private TrackVirtualManual track;
+
+        protected override bool Autoplay => autoplay;
+        private bool autoplay;
+
+        private readonly Bindable<bool> snakingIn = new Bindable<bool>();
+        private readonly Bindable<bool> snakingOut = new Bindable<bool>();
```
Comment (bdach) [inline osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs]: Not a major gripe but `BindableBool` exists.

File: osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+            track = (TrackVirtualManual)working.Track;
+            return working;
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(RulesetConfigCache configCache)
+        {
+            var config = (OsuRulesetConfigManager)configCache.GetConfigFor(Ruleset.Value.CreateInstance());
+            config.BindWith(OsuRulesetSetting.SnakingInSliders, snakingIn);
+            config.BindWith(OsuRulesetSetting.SnakingOutSliders, snakingOut);
```
Comment (bdach) [inline osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs]: We might actually want to restore these values to their initial ones at the end of each test. They are persisted to the database, which is shared with the game instance, so not restoring will actually end up interfering with other tests in the test browser or even overwriting devs' actual game settings.

File: osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+
+            addSeekStep(13500);
+            AddStep("retrieve 2nd slider repeat", () =>
+            {
+                var drawable = Player.DrawableRuleset.Playfield.AllHitObjects.Skip(1).First();
+                repeat = drawable.ChildrenOfType<Container<DrawableSliderRepeat>>().First().Children.First();
+            });
+            AddStep("Save repeat vector", () => vector = repeat.Position);
+            addSeekStep(13700);
+            AddAssert($"Repeat vector {isSame}", () => isHit ? Precision.AlmostEquals(vector.X, repeat.X, 1) && Precision.AlmostEquals(vector.Y, repeat.Y, 1) : repeat.X < vector.X && repeat.Y < vector.Y);
```
Comment (bdach) [inline osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs]: This is super hard to parse, I'd say it needs to be split into a method.

File: osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+            {
+                var body = (PlaySliderBody)slider.Body.Drawable;
+                vector = body.CurrentCurve.Last();
+            });
+            addSeekStep(1900);
+            AddAssert($"End vector {increased}", () =>
+            {
+                var body = (PlaySliderBody)slider.Body.Drawable;
+                var last = body.CurrentCurve.Last();
+                return snaking ? last.X > vector.X && last.Y > vector.Y : last == vector;
```
Comment (bdach) [inline osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs]: Why this is using strict equality and not `Precision.AlmostEquals` as above?

File: osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+        {
+            var increased = snaking ? "increased" : "is same";
+            var decreased = snaking ? "decreased" : "is same";
+
+            setSnaking(snaking);
+            addSeekStep(8800);
+            AddStep("Save end vector", () =>
+            {
+                var body = (PlaySliderBody)slider.Body.Drawable;
+                vector = body.CurrentCurve.Last();
```
Comment (bdach) [inline osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs]: This (and the `First()` pattern) repeats a lot. I'd split a method or helper property here.

**Review by bdach — DISMISSED** [review]
Test code could use some polish, it's a little repetitive even for tests. (I don't generally advocate for hard-enforcing DRY in test code but there are a few utility parts that could be factored out.)

File: osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs (hunk 6 above)
Comment (LittleEndu) [reply osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs]: body path vectors don't update at all if snaking is turned off. repeat arrow position has a chance to update depending on when in frame hit is registered.

Comment (bdach) [reply osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs]: I'd make that a comment as it's not immediately obvious.

File: osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs (hunk 2 above)
Comment (LittleEndu) [reply osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs]: I didn't know how to include strings in formatted strings (or what ever they are called in c#) so I took them outside. These are only needed for step descriptions and you could just put them in the string to begin with.

Comment (bdach) [reply osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs]: You can nest ternaries in interpolated strings but I'm not necessarily advocating for that, just better naming that will encompass both values of the strings (like `repeatVectorState` and `snakingSettingState` or something).

File: osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs (hunk 4 above)
Comment (peppy) [reply osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs]: for now this isn't a huge concern. should be fixed at a higher level.

File: osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs (hunk 2 above)
Comment (LittleEndu) [reply osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs]: I went with nested ternaries anyway. Shouldn't be that big of a deal now that the code is pretty DRY.

File: osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs (hunk 4 above)
Comment (LittleEndu) [reply osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs:57]: Is there an issue opened for this?

File: osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs (hunk 1 above)
Comment (LittleEndu) [reply osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs:192]: Should the creation of beatmap also be automatic?

**Comment by LittleEndu** [comment]
Ouch. I moved the objects in beatmap and didn't think to check the test I didn't change....

File: osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs (hunk 4 above)
Comment (peppy) [reply osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs:57]: i don't believe so

File: osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+
+        private DrawableSlider slider;
+        private DrawableSliderRepeat repeat;
+        private Vector2 savedVector;
+
+        [SetUpSteps]
+        public override void SetUpSteps() { }
+
+        [Test]
+        public void TestSnaking()
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs:68]: Putting all these tests in one `[Test]` is really hard to understand.

Split this out into `TestSnakingEnabledWithRepeats` and `TestSnakingDisabledWithRepeats`, each with [TestCase(0..3)] attributes.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by bdach** [comment]
I don't know if it's just me, but I started reading through this and now the test code complexity is through the roof. There's so much conditional control flow and bool-checking that it makes everything really opaque.

Currently have the code checked out and am trying to make it a little more readable.

**Comment by peppy** [comment]
Definitely much more can be done to make this legible. My review was just an initial pass and not a final foal.

**Comment by bdach** [comment]
After a good few hours thinking of how to decrease complexity here I ended up with [this version](https://github.com/bdach/osu/blob/7135c997466f0ed207390a8f96604b28fcf0d464/osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs). It's a bit out there in terms of implementation (`checkPositionChange` could be considered... *weird*), so I purposefully did not push it in to gather feedback on whether this version is actually legible to anyone else but me first.

**Comment by peppy** [comment]
That version already feels 10x better to read.

**Comment by bdach** [comment]
Alright, that version (with some slight retouches) is now in here too. Abstaining from further review for obvious reasons, if there's any comments to the current state of affairs drop a comment and I'll fix them up.

**Review by peppy — APPROVED** [review]
I like this
